pep-0646.rst

```diff
@@ -743,11 +753,135 @@ is available in `cpython/23527`_. A preliminary version of the version
 using the star operator, based on an early implementation of PEP 637,
 is also available at `mrahtz/cpython/pep637+646`_.
 
+Appendix: The Broader Landscape of Array Typing
+===============================================
+
+To give this PEP additional context for those particularly interested in the
+array typing use-case, here we briefly discuss design considerations
+for the venture of array typing in general.
+
+Shaped Types vs Named Axes
+--------------------------
+
+A common source of bugs in machine learning programs is incorrect selection of
+axes. For example, if we have an image stored in an array of shape 64×64x3,
+we might wish to convert to black-and-white by computing the mean over the third
+axis, ``mean(image, axis=2)``. Unfortunately, the simple typo ``axis=1`` is
+hard to spot and will produce a result that means something completely different
+(all while likely allowing the program to keep on running, resulting in a bug
+that is serious but silent).
+
+In response, some libraries have implemented so-called 'named tensors' (note:
+in this context, 'tensor' is synonymous with 'array'), in which
+axes are referred to not by index but by label - e.g. ``mean(image, axis='channels')``.
+While this ameliorates many problems, we still consider it insufficient for three
+reasons:
+
+* **Interface documentation** is still not possible with this approach. If a function should
+  *only* be willing to take array arguments that have image-like shapes, this cannot be stipulated
+  with named tensors.
+* **Static checking** of shape correctness is still not possible.
+* **Poor uptake**. Because each library much implement its own form of named tensors,
+  the burden on library maintainers in onerous - and as a result, named tensors have not
+  seen as widespread support as we might have hoped for.
```

Comment:
Interesting. I'm admittedly also uncertain about the reason for poor uptake. My impression is it's something a lot of people want and would use if available - but at least in TensorFlow it just required such a huge reworking of the data model that it was never judged a priority (though again, admittedly, now that I think about it I'm unsure where this intuition came from - could just be my own inference).

> but I think a bigger answer is that dynamic dimension names (in the data model) are fundamentally less useful in computational libraries, which do not necessarily need support arbitrary shape polymorphism

Could you expand on this? Are you saying that in deep learning libraries, the order of axes is usually clear based on the rank, because there are conventions that e.g. a 3D tensor is likely to be [width, height, channels]?

Reply:
I think a reason why named dimensions haven't worked very well in the TensorFlow ecosystem (outside of mesh tensorflow which I'll come back to later) is that in many deep net architectures most dimensions have no correspondence in reality. Pre-resnet, for example, convolutional architectures often had different values for height, width, and number of channels after every single convolution, requiring a lot of unique names to make this work. Similarly, some of these technically different dimensions happen to have the same sizes and so can be usefully mixed together, and so they do get mixed together, with reshapes inserted in many places in a model making any plan to statically relate different dimension sizes by their names conflict with their relations from their sizes. Also often in the input layers of neural networks stuff with different dimensions gets concatenated to make the input, which again is hard to express with named dimensions.

Where mesh tensorflow succeeded in naming dimensions was by giving them logical names which don't have to be related to size; you can name many dimensions "channels" or "width" and have these channels and widths partitioned the same way across your mesh even if not all channels and widths are the same across the network, in principle.

Also with transformers a lot of neural models started being written with einsum operations as the basic building blocks, which effectively uses local named dimensions, so it's not fair to say named dimensions haven't succeeded. It's just that they've been complicatedly useful in a few circumstances and not as useful for eg shape checking during model building as many of us hoped.

Reply:
> Could you expand on this? Are you saying that in deep learning libraries, the order of axes is usually clear based on the rank, because there are conventions that e.g. a 3D tensor is likely to be [width, height, channels]?

Exactly, most deep learning libraries encourage particular conventions for the semantic meanings of dimensions. If all of your tensors are either 2D [batch, channels] or 4D [batch, height, width, channels] then it's less important to explicitly name them. This is similarly true for simulation codes, though the domain specific conventions will differ. For example, in a fluid simulation code, the tensors might all be 3D [x, y, z].

More specifically:
- Named dimensions have been successful when they _provide immediate value_ like partitioning for distributed computation (Mesh TensorFlow) or self-described data analytics on arbitrary multi-dimensional data (Xarray). These are cases where the alternative is keeping track of dimension identity explicitly in some other way, so putting it into the data model saves lines of code.
- Named dimensions have not been successful (so far) for error checking/validation. The rewards are too sparse, and too delayed from when the code is written. Arguably dimension names as part of type annotations would be a better solution for this use-case, although it's still not really clear if it will be adopted.

Reply:
Interesting, I see. On the one hand the (mostly reinforcement learning) libraries I've personally been working with over the past year definitely hadn't had clear conventions like these; but on the other hand, I do remember that before I got really stuck into reinforcement learning it did tend to be clear.

My upshot from this conversation though is to be more uncertain about why named dimensions didn't take off. I'll update this bullet point to be less categorical.

Reply:
@alextp Re-reading your comment, I've just realised I didn't fully understand it.

> Pre-resnet, for example, convolutional architectures often had different values for height, width, and number of channels after every single convolution, requiring a lot of unique names to make this work.

You're saying that in one layer, `h=128, w=128`, and in the next layer, `h=64, w=64`, right?

I thought this _was_ compatible with the named tensors approach, because (at least as implemented in e.g. PyTorch) names aren't bound to sizes? E.g. the name 'h' isn't being bound to 64 - it's being bound (on a per-tensor basis) to which axis index represents the 'height-like' axis?

Reply:
Right; there are roughly two ways in which named tensors have worked. In one way (xarray) use dimension names matching dimension sizes (so if you add two tensors the dimensions with the same names are added and the dimensions with different names are broadcasted, for example) which is great for data analysis but bad for neural networks (as those have many many potentially differently sized dimensions, growing with depth). In the other way (pytorch, meshtf) names are conceptual, so you can't actually use them for sanity checking basic operations (if you add two tensors it might succeed or fail  even if those tensors are both nhwc), but you can use them in semantic ways (this function sums over the channel dimension, or this mesh layout partitions across height but not width across all layers). You can't really mix these two.

Reply:
Ahh, gotcha, thanks!